### DISCUSSION TITLE
curses invaders game in the FUZIX repository

### DISCUSSION THREAD

**Discussion by tyama501** [jbruchon/elks#2093] [Ideas]
I found curses invaders game in the FUZIX repository.

https://github.com/EtchedPixels/FUZIX/blob/master/Applications/cursesgames/invaders.c

I haven't tried porting it yet,
but we may already have most of the curses function in our curses.

**Comment by ghaerr** [comment]
It looks pretty straightforward, I would say go for it! Don't define USE_COLORS in the first pass, and our ncurses should have all of what you need. With regards to timers and keyboard functions, take a look at sl.c, ttyclock.c, matrix.c and fm.c for ideas.

**Reply by tyama501** [reply]
Hello @ghaerr 

It worked with very small modifications. (adding mvprintw, removing clrscr)
https://github.com/tyama501/elks/commit/5299a6608b227ef37e5da8d219d95f7eb8621f96

![invaders_2024-11-10_16-48-19](https://github.com/user-attachments/assets/f81bfb55-0a8a-423c-bf7a-0813b0f3b2b9)

![invaders_2024-11-10_16-49-49](https://github.com/user-attachments/assets/b5de55f0-939a-405c-a55e-8e3c5c8a5b1b)

[invaders_2024-11-10_16-50-06.webm](https://github.com/user-attachments/assets/034dd236-e510-4774-9095-826f1709e48b)

Do you have any idea to eliminate these warnings?
![image](https://github.com/user-attachments/assets/f8795009-2960-4d21-8f9c-f640d3d7383f)


Is there a way to add clrscr() to clear screen when quit?

Thank you.

**Reply by ghaerr** [reply]
Hello @tyama501,

NIce, very cool! This will be a nice game to add to the ELKS distribution :)

Thanks for the screenshots. It looks also like certain graphics characters (e.g. the Yen symbol) are being improperly displayed as the likely result of the PC-98 codepage/character set being different than IBM's CP 437. So you might want to look at which characters these are and replace them for the PC-98 version. If `invaders` is run on QEMU it will show the CP 437 glyphs.

> Is there a way to add clrscr() to clear screen when quit?

You should be able to clear the screen by renaming it to erase(), it looks to be the same as erase() in curses.c.

> Do you have any idea to eliminate these warnings?

Try adding `int abs(int);` for a definition of abs, and change `char *fmt` to `const char *fmt` in the declaration of mvprintw.

Thank you!

**Reply by tyama501** [reply]
Thank you. I didn't notice that.
Historically Japanese fonts uses Yen as the backslash. I think there is a reason but I forgot what it is.
Maybe I will leave it as is.

Isn't the abs declared in the stdlib or something?

Also, mvprintw in the curses is normally not using const right?
(Well, maybe it is ok to change as const for now)

**Reply by ghaerr** [reply]
> Historically Japanese fonts uses Yen as the backslash.

I see now - the missile guns would graphically look better with a backslash on the top right, and PC-98's character set replaces \ with the Yen symbol. You might think of replacing it with something more graphically appealing if \ is not available.

> Isn't the abs declared in the stdlib or something?

No, although I've noticed it being missing a few times now. It seems abs is being picked up by GCC's internal library. I'll add `abs` in a PR to fix this.

> Also, mvprintw in the curses is normally not using const right?
(Well, maybe it is ok to change as const for now)

Go ahead and change the declaration and see whether other TUI programs issue warnings on compilation. It is correct to use const as the format string is a const pointer.

**Reply by ghaerr** [reply]
I added some comments on your commit to your local ELKS repository.

**Reply by tyama501** [reply]
Thank you.
I will modify it.

deleteln is actually not used since clean is zero in the caller.
paintAlienRow(int row, int clean)

**Reply by toncho11** [reply]
Wow. Great job @tyama501 !

**Reply by tyama501** [reply]
Hi @toncho11 
Thank you.

Hi @ghaerr 

I have cleaned the code.
https://github.com/tyama501/elks/commit/93c79596457a6dc0ab5630600b89f01a266c2c99

I also added the key instruction in the title.
https://github.com/tyama501/elks/commit/13649fe1115e8706e090058c4a5f02761efe1a79

I will make a first PR.

**Reply by tyama501** [reply]
https://github.com/ghaerr/elks/discussions/2093
